### PR TITLE
chore: relative url for notification links

### PR DESCRIPTION
**Pull request by Dhanus3133**
Fixes: #249 

image

### CODE REVIEW

File: openwisp_notifications/base/models.py [hunk 1]
```diff
@@ -84,22 +84,19 @@ def _get_related_object_url(self, field):
...
-    @property
-    def action_url(self):
-        return self._get_related_object_url(field='action_object')
+    def action_url(self, absolute_url=False):
+        return self._get_related_object_url(field='action_object', absolute_url=absolute_url)
 
-    @property
-    def target_url(self):
-        return self._get_related_object_url(field='target')
+    def target_url(self, absolute_url=False):
```
Comment (nemesifier) [inline openwisp_notifications/base/models.py:98]: If we turn these properties into methods which support passing arguments we have to change their name to `get_{property_name}`, for backward compatibility we can keep the properties and internally call the new method which passes `absolute_url=True`.

**Review by nemesifier — CHANGES_REQUESTED** [review]
Thank you @Dhanus3133, it looks good.

Here's a few things to improve, see below.

The REST API is returning relative URLs as shown in the following screenshot:

Screenshot from 2024-03-07 16-29-40

That's unintended, we only want this new behavior in the admin widget, the API may be hosted on a different domain and hence must return the full URL.

Please also add a new unit test which tests the new methods that allow returning the relative URLs.

Watch out also from build failures because of failed QA checks, please read the build output to find out how to run qa checks locally and how to fix common formatting and commit message format issues.

**Comment by Dhanus3133** [comment]
Updated with just the required notification widget changes, although the QA checks for the CSS files are failing for some reason.

**Review by nemesifier — CHANGES_REQUESTED** [review]
Works as expected now, thanks!
Take a look at the comments below for some minor code improvements.

If you can add a selenium browser test for this it would be great, let me know!

**Comment by Dhanus3133** [comment]
@nemesifier, I tried adding Selenium tests, but when the notification tab is clicked, it just loads. When I look into the console, it says it cannot find WS and returns 404 in the network.
image
Could you help me here? Thanks.

**Comment by nemesifier** [comment]
> it just loads

Did you mean "it doesn't load"?

**Comment by Dhanus3133** [comment]
> Did you mean "it doesn't load"?

I meant the WebSocket connection is not found, so in the front end, it just waits, causing the spinning UI.

https://github.com/openwisp/openwisp-notifications/assets/43700516/1c21bb68-f2ec-4c4c-9fb5-da85fe4b112e

**Review by nemesifier — CHANGES_REQUESTED** [review]
The base test class `StaticLiveServerTestCase` does not support websockets.

Here's a rough outline of how we might structure these tests:

```python
from django.contrib.staticfiles.testing import StaticLiveServerTestCase
from channels.testing import ChannelsLiveServerTestCase

class SeleniumTestNotifications(StaticLiveServerTestCase, ChannelsLiveServerTestCase):
    # ... etc ...
```

Please consults these links for more information:

- [ChannelsLiveServerTestCase docs](https://channels.readthedocs.io/en/latest/topics/testing.html) 
- [StaticLiveServerTestCase docs](https://docs.djangoproject.com/en/5.0/ref/contrib/staticfiles/#specialized-test-case-to-support-live-testing)

**Review by nemesifier — COMMENTED** [review]
@Dhanus3133 I missed the latest updates on this, will review asap.

File: openwisp_notifications/tests/test_notifications.py [hunk 2]
```diff
@@ -927,3 +932,48 @@ def test_notification_cache_update(self):
         self.assertEqual(notification.target.username, 'new operator name')
         # Done for populating cache
         self.assertEqual(operator_cache.username, 'new operator name')
+
+
+class SeleniumTestNotifications(
```
Comment (nemesifier) [inline openwisp_notifications/tests/test_notifications.py:937]: I would move this to a new dedicated test file, it's possible that over time we'll be adding more tests for the widget.

File: requirements-test.txt [hunk 3]
```diff
@@ -5,3 +5,4 @@ channels_redis~=4.1.0
 pytest-asyncio~=0.21.0
 pytest-django~=4.5.0
 freezegun~=1.2.2
+selenium~=4.18.1
```
Comment (nemesifier) [inline requirements-test.txt:8]: We already depend on selenium in openwisp-utils[selenium], we may want to look if we can avoid redefining this here: https://github.com/openwisp/openwisp-utils/blob/master/setup.py#L67

File: tests/openwisp2/settings.py [hunk 4]
```diff
@@ -14,6 +14,9 @@
     'default': {
         'ENGINE': 'django.db.backends.sqlite3',
         'NAME': os.path.join(BASE_DIR, 'openwisp_notifications.db'),
+        "TEST": {
+            "NAME": os.path.join(BASE_DIR, "openwisp_notifications_test.db"),
+        },
```
Comment (nemesifier) [inline tests/openwisp2/settings.py:19]: Are you sure this is really needed?

**Review by nemesifier — COMMENTED** [review]
Great work @Dhanus3133, I left some comments below, I am doing some quick tests locally  to verify some of my comments.

**Review by nemesifier — APPROVED** [review]
@Dhanus3133 I pushed some improvements, feel free to review the changes. Once the build passes I will merge. :+1:

**Pull request merged**